File: BasicRxJavaSample/app/src/main/java/com/example/android/observability/demoTest/ImitateViewModelFactory.java
package com.example.android.observability.demoTest;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.android.observability.ui.UserViewModel;

public class ImitateViewModelFactory implements ViewModelProvider.Factory {
    private ImitateUserDataSource mUserDataSource;

    public ImitateViewModelFactory(ImitateUserDataSource userDataSource) {
        mUserDataSource = userDataSource;
    }
    /**
     * ；泛型好用：可以灵活扩展
     *
     * @param modelClass
     * @param <T>
     * @return
     */
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(ImitateUserViewModel.class)) {
            return (T) new ImitateUserViewModel(mUserDataSource);
        } else if (modelClass.isAssignableFrom(UserViewModel.class)) {
//            return (T) new UserViewModel()
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: BasicRxJavaSampleKotlin/app/src/main/java/com/example/android/observability/imitate/ui/ImitateViewModelFactory.kt
package com.example.android.observability.imitate.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.android.observability.imitate.persistance.ImitateLocalDataSource

class ImitateViewModelFactory(val dataSource: ImitateLocalDataSource) : ViewModelProvider.Factory{

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(ImitateUserViewModel::class.java)) {
            return ImitateUserViewModel(dataSource) as T
        }
        throw IllegalAccessException("unKnown viewmodel class")
    }
}